File: drills/python/solutions/nc120_min_cost_connect_points__kruskal.py
def minCostConnectPoints(points):
    # Kruskal's algorithm: every edge, cheapest first, kept only when it joins
    # two pieces that are not already connected. Union-find is what makes that
    # test cheap. The trade against Prim's is the sort -- O(n^2 log n) edges
    # here against Prim's O(n^2) -- but Kruskal never needs the points
    # themselves, only the edge list, so it is the one that generalises to a
    # sparse graph.
    edges = []
    for i in range(len(points)):
        for j in range(i + 1, len(points)):
            edges.append((distance(points[i], points[j]), i, j))
    edges.sort()

    parents = {}

    def find(node):
        while parents.get(node, node) != node:
            node = parents[node]
        return node

    total = 0
    for cost, i, j in edges:
        rootI, rootJ = find(i), find(j)
        if rootI != rootJ:
            parents[rootI] = rootJ
            total += cost

    return total


def distance(a, b):
    return abs(a[0] - b[0]) + abs(a[1] - b[1])

File: drills/elixir/solutions/nc120_min_cost_connect_points__kruskal.exs
defmodule Solution do
  # Kruskal's algorithm: every edge, cheapest first, kept only when it joins two
  # pieces that are not already connected. Union-find is what makes that test
  # cheap. The trade against Prim's is the sort -- O(n^2 log n) edges here
  # against Prim's O(n^2) -- but Kruskal never needs the points themselves, only
  # the edge list, so it is the one that generalises to a sparse graph.
  def min_cost_connect_points(points) do
    indexed = Enum.with_index(points)

    edges =
      for {a, i} <- indexed, {b, j} <- indexed, i < j do
        {distance(a, b), i, j}
      end
      |> Enum.sort()

    {_parents, total} =
      Enum.reduce(edges, {%{}, 0}, fn {cost, i, j}, {parents, total} ->
        root_i = find(parents, i)
        root_j = find(parents, j)

        if root_i == root_j,
          do: {parents, total},
          else: {Map.put(parents, root_i, root_j), total + cost}
      end)

    total
  end

  defp find(parents, node) do
    case Map.get(parents, node, node) do
      ^node -> node
      parent -> find(parents, parent)
    end
  end

  defp distance([ax, ay], [bx, by]), do: abs(ax - bx) + abs(ay - by)
end
